File: priv/parsers/ruby/parser.rb
#!/usr/bin/env ruby
# encoding: UTF-8
# frozen_string_literal: true

# Ruby AST Parser for RSOLV RFC-031
# Uses Ruby's built-in AST parser to parse Ruby code and returns AST in JSON format via stdin/stdout

# Set encoding for STDIN/STDOUT to UTF-8
Encoding.default_external = Encoding::UTF_8
Encoding.default_internal = Encoding::UTF_8
STDIN.set_encoding(Encoding::UTF_8)
STDOUT.set_encoding(Encoding::UTF_8)

# Try to use bundler if available, otherwise just require gems directly
begin
  Dir.chdir(File.dirname(__FILE__))
  require 'bundler/setup'
rescue LoadError
  # Bundler not available, gems should be installed globally
end

require 'json'
# Use Prism directly for better error detection and unicode handling
require 'prism'
require 'timeout'

# Set up signal handler for timeout
Signal.trap('ALRM') do
  error_response = {
    'status' => 'error',
    'error' => {
      'type' => 'TimeoutError',
      'message' => 'Parser timeout after 30 seconds'
    }
  }
  puts JSON.generate(error_response)
  STDOUT.flush
  exit(1)
end

def node_to_dict(node)
  # Convert Prism::Node to hash format with circular reference handling
  return nil if node.nil?

  # Handle Prism nodes
  if node.respond_to?(:type)
    # Strip _node suffix from Prism types for compatibility with old Parser gem format
    type_str = node.type.to_s.sub(/_node$/, '')
    result = {
      'type' => type_str
    }

    # Add location info if available
    if node.respond_to?(:location) && node.location
      loc = node.location
      result['_loc'] = {
        'start' => {
          'line' => loc.start_line,
          'column' => loc.start_column
        },
        'end' => {
          'line' => loc.end_line,
          'column' => loc.end_column
        }
      }
      result['_start'] = loc.start_offset
      result['_end'] = loc.end_offset
    end

    # Process child nodes
    children = []
    if node.respond_to?(:child_nodes)
      children = node.child_nodes.compact.map { |child| node_to_dict(child) }
    end

    result['children'] = children unless children.empty?
    result
  else
    # Return primitive values as-is
    node
  end
end

def find_security_patterns(ast)
  # Extract security-relevant patterns from Ruby AST (Prism)
  patterns = []

  def traverse_node(node, patterns)
    return unless node && node.respond_to?(:type)

    case node.type
    when :call_node
      # Method calls - check for dangerous methods
      method_name = node.respond_to?(:name) ? node.name : nil
      
      if method_name
        method_str = method_name.to_s
        
        # Common dangerous methods
        dangerous_methods = %w[
          eval exec system ` shell_escape
          instance_eval class_eval module_eval
          const_get const_set remove_const
          send __send__ public_send
          html_safe raw
        ]
        
        loc = node.location if node.respond_to?(:location)
        line = loc ? loc.start_line : 0
        column = loc ? loc.start_column : 0

        if dangerous_methods.include?(method_str)
          patterns << {
            'type' => 'dangerous_method',
            'method' => method_str,
            'line' => line,
            'column' => column
          }
        end

        # SQL injection patterns
        if method_str.match?(/^(find|where|execute|query)$/)
          patterns << {
            'type' => 'potential_sql_injection',
            'method' => method_str,
            'line' => line,
            'column' => column
          }
        end

        # HTML output methods
        if method_str.match?(/^(html_safe|raw)$/)
          patterns << {
            'type' => 'html_output',
            'method' => method_str,
            'line' => line,
            'column' => column
          }
        end
      end

    when :interpolated_string_node, :interpolated_x_string_node
      # String interpolation and command execution
      loc = node.location if node.respond_to?(:location)
      patterns << {
        'type' => node.type == :interpolated_string_node ? 'string_interpolation' : 'command_execution',
        'line' => loc ? loc.start_line : 0,
        'column' => loc ? loc.start_column : 0
      }

    when :constant_read_node
      # Constant access - check for dangerous constants
      const_name = node.respond_to?(:name) ? node.name : nil
      if const_name && const_name.to_s.match?(/^(Kernel|Object|BasicObject)$/)
        loc = node.location if node.respond_to?(:location)
        patterns << {
          'type' => 'dangerous_constant',
          'constant' => const_name.to_s,
          'line' => loc ? loc.start_line : 0,
          'column' => loc ? loc.start_column : 0
        }
      end
    end

    # Recursively traverse children
    if node.respond_to?(:child_nodes)
      node.child_nodes.compact.each do |child|
        traverse_node(child, patterns)
      end
    end
  end
  
  traverse_node(ast, patterns)
  patterns
end

def main
  # Main parser loop - reads JSON requests from stdin, writes responses to stdout
  STDOUT.sync = true
  
  while line = STDIN.gets
    begin
      # Set 30-second timeout for each request
      Timeout::timeout(30) do
        request = JSON.parse(line.strip)
        request_id = request['id'] || 'unknown'
        command = request['command'] || ''
        action = request['action'] || command # Support both formats
        
        if action == 'HEALTH_CHECK'
          response = {
            'id' => request_id,
            'result' => 'ok'
          }
        elsif !action.empty? && action != 'parse'
          # Handle command-based interface (for compatibility with PortWorker)
          start_time = Time.now
          code = command # Treat command as code to parse
          options = request['options'] || {}
          filename = request['filename'] || '<string>'

          # First, check for syntax errors using Prism directly
          prism_result = Prism.parse(code, filepath: filename)
          if !prism_result.success? && !prism_result.errors.empty?
            # Syntax errors detected - return error response
            first_error = prism_result.errors.first
            raise_error = {
              'id' => request_id,
              'status' => 'error',
              'success' => false,
              'error' => {
                'type' => 'SyntaxError',
                'message' => first_error.message,
                'line' => first_error.location.start_line,
                'column' => first_error.location.start_column
              }
            }
            puts JSON.generate(raise_error)
            STDOUT.flush
            next
          end

          # Parse the Ruby code using Prism AST
          ast = prism_result.value
          
          # Convert AST to dictionary
          ast_dict = node_to_dict(ast)
          
          # Extract security patterns if requested
          security_patterns = []
          if options['include_security_patterns'] != false
            security_patterns = find_security_patterns(ast)
          end
          
          parse_time_ms = ((Time.now - start_time) * 1000).to_i
          
          # Count nodes by traversing AST
          node_count = 0
          count_nodes = lambda do |node|
            if node && node.respond_to?(:child_nodes)
              node_count += 1
              node.child_nodes.compact.each { |child| count_nodes.call(child) }
            elsif node.is_a?(Array)
              node.compact.each { |item| count_nodes.call(item) }
            end
          end
          count_nodes.call(ast)
          
          response = {
            'id' => request_id,
            'status' => 'success',
            'success' => true,
            'ast' => ast_dict,
            'security_patterns' => security_patterns,
            'metadata' => {
              'parser_version' => '1.0.0',
              'language' => 'ruby',
              'language_version' => RUBY_VERSION,
              'parse_time_ms' => parse_time_ms,
              'ast_node_count' => node_count
            }
          }
        elsif action == 'parse' || action.empty?
          start_time = Time.now
          code = request['code'] || ''
          options = request['options'] || {}
          filename = request['filename'] || '<string>'

          # First, check for syntax errors using Prism directly
          prism_result = Prism.parse(code, filepath: filename)
          if !prism_result.success? && !prism_result.errors.empty?
            # Syntax errors detected - return error response
            first_error = prism_result.errors.first
            raise_error = {
              'id' => request_id,
              'status' => 'error',
              'success' => false,
              'error' => {
                'type' => 'SyntaxError',
                'message' => first_error.message,
                'line' => first_error.location.start_line,
                'column' => first_error.location.start_column
              }
            }
            puts JSON.generate(raise_error)
            STDOUT.flush
            next
          end

          # Parse the Ruby code using Prism AST
          ast = prism_result.value
          
          # Convert AST to dictionary
          ast_dict = node_to_dict(ast)
          
          # Extract security patterns if requested
          security_patterns = []
          if options['include_security_patterns'] != false
            security_patterns = find_security_patterns(ast)
          end
          
          parse_time_ms = ((Time.now - start_time) * 1000).to_i
          
          # Count nodes
          node_count = 0
          count_nodes = lambda do |node|
            if node && node.respond_to?(:child_nodes)
              node_count += 1
              node.child_nodes.compact.each { |child| count_nodes.call(child) }
            elsif node.is_a?(Array)
              node.compact.each { |item| count_nodes.call(item) }
            end
          end
          count_nodes.call(ast)
          
          response = {
            'id' => request_id,
            'status' => 'success',
            'ast' => ast_dict,
            'security_patterns' => security_patterns,
            'metadata' => {
              'parser_version' => '1.0.0',
              'language' => 'ruby',
              'language_version' => RUBY_VERSION,
              'parse_time_ms' => parse_time_ms,
              'ast_node_count' => node_count
            }
          }
        else
          response = {
            'id' => request_id,
            'status' => 'error',
            'error' => {
              'type' => 'InvalidAction',
              'message' => "Unknown action: #{action}"
            }
          }
        end
        
        puts JSON.generate(response)
        STDOUT.flush
      end
      
    rescue Timeout::Error
      response = {
        'id' => (request && request['id']) || 'unknown',
        'status' => 'error',
        'success' => false,
        'error' => {
          'type' => 'TimeoutError',
          'message' => 'Parser timeout after 30 seconds'
        }
      }
      puts JSON.generate(response)
      STDOUT.flush
      
    rescue JSON::ParserError => e
      response = {
        'id' => 'unknown',
        'status' => 'error',
        'success' => false,
        'error' => {
          'type' => 'JSONParserError',
          'message' => "Invalid JSON: #{e.message}"
        }
      }
      puts JSON.generate(response)
      STDOUT.flush
      
    rescue => e
      request_id = 'unknown'
      begin
        request = JSON.parse(line.strip) if line
        request_id = request['id'] if request
      rescue
        # Ignore errors for request ID extraction
      end
      
      response = {
        'id' => request_id,
        'status' => 'error',
        'success' => false,
        'error' => {
          'type' => e.class.name,
          'message' => e.message
        }
      }
      puts JSON.generate(response)
      STDOUT.flush
    end
  end
end

if __FILE__ == $0
  main
end

File: priv/parsers/php/parser.php
<?php
/**
 * PHP AST Parser for RSOLV RFC-031
 * Uses nikic/php-parser to parse PHP code and return AST in JSON format via stdin/stdout
 */

// Set error reporting
error_reporting(E_ALL & ~E_NOTICE);
ini_set('display_errors', 0);

// Change to parser directory to find composer autoload
chdir(__DIR__);

// Include composer autoload
if (file_exists('vendor/autoload.php')) {
    require_once 'vendor/autoload.php';
} else {
    // Fallback error response
    $error = [
        'id' => 'unknown',
        'status' => 'error',
        'success' => false,
        'error' => [
            'type' => 'ComposerError',
            'message' => 'Composer dependencies not installed. Run: composer install'
        ]
    ];
    echo json_encode($error) . "\n";
    exit(1);
}

use PhpParser\Error;
use PhpParser\NodeDumper;
use PhpParser\ParserFactory;
use PhpParser\PrettyPrinter;
use PhpParser\Node;

class SecurityPatternFinder
{
    private $patterns = [];

    public function findPatterns($ast)
    {
        $this->patterns = [];
        $this->traverseNode($ast);
        return $this->patterns;
    }

    private function traverseNode($node)
    {
        if ($node instanceof Node) {
            $this->checkNodeForSecurityPatterns($node);
            
            // Traverse all child nodes
            foreach ($node->getSubNodeNames() as $name) {
                $child = $node->$name;
                if (is_array($child)) {
                    foreach ($child as $item) {
                        $this->traverseNode($item);
                    }
                } else {
                    $this->traverseNode($child);
                }
            }
        }
    }

    private function checkNodeForSecurityPatterns($node)
    {
        // Function/method calls
        if ($node instanceof Node\Expr\FuncCall) {
            $funcName = null;
            if ($node->name instanceof Node\Name) {
                $funcName = $node->name->toString();
            } elseif ($node->name instanceof Node\Expr\Variable && $node->name->name === 'this') {
                // Handle $this->method() calls
                return;
            }
            
            if ($funcName) {
                // Check for dangerous functions
                $dangerousFunctions = [
                    'eval', 'exec', 'system', 'shell_exec', 'passthru',
                    'popen', 'proc_open', 'file_get_contents', 'file_put_contents',
                    'fopen', 'fwrite', 'include', 'include_once', 'require', 'require_once',
                    'unserialize', 'preg_replace', 'create_function',
                    'call_user_func', 'call_user_func_array'
                ];
                
                if (in_array(strtolower($funcName), $dangerousFunctions)) {
                    $this->patterns[] = [
                        'type' => 'dangerous_function',
                        'function' => $funcName,
                        'line' => $node->getLine() ?? 0,
                        'column' => $node->getStartFilePos() ?? 0
                    ];
                }
                
                // Check for SQL-related functions
                $sqlFunctions = ['mysql_query', 'mysqli_query', 'pg_query', 'sqlite_query'];
                if (in_array(strtolower($funcName), $sqlFunctions)) {
                    $this->patterns[] = [
                        'type' => 'potential_sql_injection',
                        'function' => $funcName,
                        'line' => $node->getLine() ?? 0,
                        'column' => $node->getStartFilePos() ?? 0
                    ];
                }
            }
        }
        
        // Method calls
        if ($node instanceof Node\Expr\MethodCall) {
            if ($node->name instanceof Node\Identifier) {
                $methodName = $node->name->toString();
                
                // Check for dangerous methods
                $dangerousMethods = ['query', 'exec', 'prepare', 'execute'];
                if (in_array(strtolower($methodName), $dangerousMethods)) {
                    $this->patterns[] = [
                        'type' => 'potential_sql_injection',
                        'method' => $methodName,
                        'line' => $node->getLine() ?? 0,
                        'column' => $node->getStartFilePos() ?? 0
                    ];
                }
            }
        }
        
        // String concatenation that might lead to injection
        if ($node instanceof Node\Expr\BinaryOp\Concat) {
            $this->patterns[] = [
                'type' => 'string_concatenation',
                'line' => $node->getLine() ?? 0,
                'column' => $node->getStartFilePos() ?? 0
            ];
        }
        
        // Variable variables ($$var)
        if ($node instanceof Node\Expr\Variable && $node->name instanceof Node\Expr) {
            $this->patterns[] = [
                'type' => 'variable_variable',
                'line' => $node->getLine() ?? 0,
                'column' => $node->getStartFilePos() ?? 0
            ];
        }
    }
}

function nodeToArray($node)
{
    if ($node === null) {
        return null;
    }
    
    if (is_array($node)) {
        $result = [];
        foreach ($node as $key => $value) {
            $result[$key] = nodeToArray($value);
        }
        return $result;
    }
    
    if (!($node instanceof Node)) {
        return $node;
    }
    
    $result = [
        'type' => $node->getType()
    ];
    
    // Add location information
    if ($node->getLine() !== -1) {
        $result['_loc'] = [
            'start' => [
                'line' => $node->getLine(),
                'column' => $node->getStartFilePos() ?? 0
            ],
            'end' => [
                'line' => $node->getEndLine() ?? $node->getLine(),
                'column' => $node->getEndFilePos() ?? ($node->getStartFilePos() ?? 0)
            ]
        ];
        
        $result['_start'] = $node->getStartFilePos() ?? 0;
        $result['_end'] = $node->getEndFilePos() ?? 0;
    }
    
    // Add children
    $children = [];
    foreach ($node->getSubNodeNames() as $name) {
        $child = $node->$name;
        $children[$name] = nodeToArray($child);
    }
    
    if (!empty($children)) {
        $result['children'] = $children;
    }
    
    return $result;
}

function processRequest($jsonInput)
{
    try {
        $request = json_decode($jsonInput, true);
        if (!$request) {
            throw new Exception('Invalid JSON input');
        }
        
        $requestId = $request['id'] ?? 'unknown';
        $command = $request['command'] ?? '';
        $action = $request['action'] ?? $command;
        
        if ($action === 'HEALTH_CHECK') {
            $response = [
                'id' => $requestId,
                'result' => 'ok'
            ];
            echo json_encode($response) . "\n";
            return;
        }
        
        $startTime = microtime(true);
        $code = '';
        $options = $request['options'] ?? [];
        $filename = $request['filename'] ?? '<string>';
        
        if ($action !== 'parse' && !empty($action)) {
            // Command-based interface
            $code = $command;
        } else {
            // Standard parse interface
            $code = $request['code'] ?? '';
        }
        
        // Parse PHP code
        $parser = (new ParserFactory())->create(ParserFactory::PREFER_PHP7);
        $ast = $parser->parse($code);
        
        // Convert AST to array
        $astArray = nodeToArray($ast);
        
        // Extract security patterns
        $securityPatterns = [];
        $includeSecurityPatterns = !isset($options['include_security_patterns']) || 
                                  $options['include_security_patterns'] !== false;
        
        if ($includeSecurityPatterns) {
            $finder = new SecurityPatternFinder();
            $securityPatterns = $finder->findPatterns($ast);
        }
        
        $parseTime = (microtime(true) - $startTime) * 1000;
        
        // Count nodes
        $nodeCount = countNodes($ast);
        
        $response = [
            'id' => $requestId,
            'status' => 'success',
            'success' => true,
            'ast' => $astArray,
            'security_patterns' => $securityPatterns,
            'metadata' => [
                'parser_version' => '1.0.0',
                'language' => 'php',
                'language_version' => phpversion(),
                'parse_time_ms' => (int)$parseTime,
                'ast_node_count' => $nodeCount
            ]
        ];
        
        echo json_encode($response) . "\n";
        
    } catch (Error $e) {
        $requestId = $request['id'] ?? 'unknown';
        $response = [
            'id' => $requestId,
            'status' => 'error',
            'success' => false,
            'error' => [
                'type' => 'SyntaxError',
                'message' => $e->getMessage(),
                'line' => method_exists($e, 'getStartLine') ? $e->getStartLine() : 0,
                'column' => 0
            ]
        ];
        echo json_encode($response) . "\n";
        
    } catch (Exception $e) {
        $requestId = isset($request) ? ($request['id'] ?? 'unknown') : 'unknown';
        $response = [
            'id' => $requestId,
            'status' => 'error',
            'success' => false,
            'error' => [
                'type' => get_class($e),
                'message' => $e->getMessage()
            ]
        ];
        echo json_encode($response) . "\n";
    }
}

function countNodes($node)
{
    if (!($node instanceof Node)) {
        return 0;
    }
    
    $count = 1; // Count current node
    
    foreach ($node->getSubNodeNames() as $name) {
        $child = $node->$name;
        if (is_array($child)) {
            foreach ($child as $item) {
                $count += countNodes($item);
            }
        } else {
            $count += countNodes($child);
        }
    }
    
    return $count;
}

// Main execution
function main()
{
    // Set timeout signal handler
    if (function_exists('pcntl_signal')) {
        pcntl_signal(SIGALRM, function() {
            $error = [
                'id' => 'unknown',
                'status' => 'error',
                'success' => false,
                'error' => [
                    'type' => 'TimeoutError',
                    'message' => 'Parser timeout after 30 seconds'
                ]
            ];
            echo json_encode($error) . "\n";
            exit(1);
        });
        pcntl_alarm(30); // 30 second timeout
    }
    
    // Read from stdin
    while (($line = fgets(STDIN)) !== false) {
        $line = trim($line);
        if (!empty($line)) {
            processRequest($line);
            flush();
        }
    }
}

if (php_sapi_name() === 'cli') {
    main();
}
?>